### PR TITLE
fix(auth): Restrict deletions for session limit operations

### PR DESCRIPTION
**Pull request by dschom**
## Because

- We have reason to believe heavy handed token pruning operations where having adverse side effects on other queries.

## This pull request
- Limits the number of accounts pruned that can be pruned per run.
- Limits the number of session tokens pruned per account per run.

## Issue that this pull request solves

Closes: #FXA-5797

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

### CODE REVIEW

**Review by chenba — APPROVED** [review]
Approving this as it does what it says on the tin.  However, I'm uncertain that it will address the performance issue since that _might_ be from subquery counting the sessions per uid.

File: packages/db-migrations/databases/fxa/patches/patch-134-135.sql [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+-- This migration adds a routine for removing sessions on accounts that have
+-- over maxSessions. Essentially, per account tokens above maxSessions
+-- will be purged. Older tokens will be removed first. Associated unverified
+-- tokens and devices will also be removed.
+--   maxSessions - Max number of sessions allowed per account
+--   accountLimit - Max number of accouts to process at a time
+--   deletionLimit - Max number of session tokens to delete at a time
+--   accountsOverLimit - Number of accounts having more than the allowed session token limit
+--   totalDeletions - Total number of rows deleted, which includes session token rows, unverified session token rows and device rows.
+CREATE PROCEDURE `limitSessions_2` (
```
[the remark this answers is not in the document]
Comment (dschom) [reply packages/db-migrations/databases/fxa/patches/patch-134-135.sql:15]: We can. Do we usually drop old sprocs?

[the remark this answers is not in the document]
Comment (chenba) [reply packages/db-migrations/databases/fxa/patches/patch-134-135.sql:15]: We haven't been consistent with it.

[the remark this answers is not in the document]
Comment (dschom) [reply packages/db-migrations/databases/fxa/patches/patch-134-135.sql:15]: It seems like the most common thing to do is to only drop the procedure in a downgrade patch. This makes sense to me, and is what I am used to seeing. It also maintains better backwards compatibility and can avoid issues during rollout.

[the remark this answers is not in the document]
Comment (chenba) [reply packages/db-migrations/databases/fxa/patches/patch-134-135.sql:15]: > maintains better backwards compatibility and can avoid issues during rollout.

Oh yeah, I'm not suggesting we drop the old sproc in this patch.

File: packages/db-migrations/databases/fxa/patches/patch-134-135.sql [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    DROP TEMPORARY TABLE IF EXISTS tokenCounts;
+    CREATE TEMPORARY TABLE tokenCounts
+      SELECT uid
+      FROM (
+        SELECT uid, COUNT(tokenId) AS tokenCount
+        FROM sessionTokens
+        GROUP BY uid
+      ) AS counts
+      WHERE tokenCount > maxSessions
+      ORDER BY counts.tokenCount DESC
```
Comment (dschom) [inline packages/db-migrations/databases/fxa/patches/patch-134-135.sql:51]: @chenba I just added this to make sure we are really addressing the accounts with the highest session counts first. This `ORDER BY` probably adds some expense though...

File: packages/fxa-auth-server/test/scripts/prune-tokens.js [hunk 3]
```diff
@@ -324,43 +327,83 @@ describe('scripts/prune-tokens', () => {
...
-          size - 1
-        }' `,
+        `NODE_ENV=dev node -r esbuild-register scripts/prune-tokens.ts ${args}`,
         {
           cwd,
           shell: '/bin/bash',
         }
       );
 
+      console.log('!!! stderr', stderr);
```
Comment (pdehaan) [inline packages/fxa-auth-server/test/scripts/prune-tokens.js:340]: Q: Did you want this in, or was this debugging code?

**Comment by dschom** [comment]
@chenba I am going to close and reopen this pull request. After feedback provided, I did another iteration. There's enough differences that a fresh review is needed. A new PR will be issued shortly.

**Pull request closed without merging**